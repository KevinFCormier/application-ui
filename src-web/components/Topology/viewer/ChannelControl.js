/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import R from 'ramda'
import PropTypes from 'prop-types'
import { DropdownV2 } from 'carbon-components-react'
import msgs from '../../../../nls/platform.properties'
import _ from 'lodash'

class ChannelControl extends React.Component {
  static propTypes = {
    channelControl: PropTypes.shape({
      allChannels: PropTypes.array,
      activeChannel: PropTypes.string,
      isChangingChannel: PropTypes.bool,
      changeTheChannel: PropTypes.func
    }),
    locale: PropTypes.string
  };

  constructor(props) {
    super(props)
    this.state = {}
  }

  componentDidMount() {
    this.syncHeight = this.syncHeight.bind(this)
    window.addEventListener('resize', this.syncHeight)
    this.syncHeight()

    if (this.scrollIntoViewChn) {
      setTimeout(() => {
        this.scrollIntoViewChn.scrollIntoView(false)
      }, 100)
    }
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.syncHeight)
  }

  // sync scrollbar height with diagram height
  syncHeight() {
    var div = document.getElementById('diagram-viewer-container-container')
    if (div) {
      this.setState({ diagramHeight: div.getBoundingClientRect().height })
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      !_.isEqual(this.props.channelControl, nextProps.channelControl) ||
      this.state.diagramHeight !== nextState.diagramHeight
    )
  }

  handleSubscriptionChange = e => {
    const channel = e.selectedItem.chn
    this.changeSubscriptionChannels(channel)
  };

  selectChannelByNumber(channelNb) {
    const allChannels = R.pathOr([], ['channelControl', 'allChannels'])(
      this.props
    )

    const changeToChannel =
      allChannels.length >= channelNb ? allChannels[channelNb - 1] : null
    this.changeSubscriptionChannels(changeToChannel)
  }

  handlePageChanged = e => {
    const selectedValue = e.target.value
    this.selectChannelByNumber(selectedValue)
  };

  handlePageClick = e => {
    const allChannels = R.pathOr([], ['channelControl', 'allChannels'])(
      this.props
    )
    const activeChannel = R.pathOr(null, ['channelControl', 'activeChannel'])(
      this.props
    )
    const selectedChannelIndex =
      activeChannel && allChannels ? allChannels.indexOf(activeChannel) + 1 : 1
    switch (e.target.id) {
    case 'p1': {
      //move to the first channel
      this.selectChannelByNumber(1)
      break
    }
    case 'p2': {
      //move one channel down
      if (selectedChannelIndex > 0) {
        this.selectChannelByNumber(selectedChannelIndex - 1)
      }
      break
    }
    case 'p3': {
      //move one channel up
      if (selectedChannelIndex < allChannels.length) {
        this.selectChannelByNumber(selectedChannelIndex + 1)
      }

      break
    }
    case 'p4': {
      //up to the last channel
      this.selectChannelByNumber(allChannels.length)
      break
    }
    default:
      break
    }
  };

  render() {
    const { channelControl = {}, locale } = this.props
    const { allChannels } = channelControl

    if (allChannels) {
      let { activeChannel } = channelControl
      const { fetchChannel } = this.state
      activeChannel = fetchChannel || activeChannel

      // determine if there are subchannels
      const channelMap = {}
      allChannels.forEach(chnl => {
        const [chn, beg, end] = chnl.split('///')
        const splitChn = /(.*)\/(.*)\/\/(.*)\/(.*)/.exec(chn)
        if (splitChn && splitChn.length === 5) {
          let data = channelMap[chn]
          if (!data) {
            data = channelMap[chn] = { chnl, splitChn, subchannels: [] }
          }
          if (beg && end) {
            data.subchannels.push({ chnl, beg, end })
          }
        }
      })
      // determine displayed channels
      const displayChannels = []
      const showMainChannel = Object.keys(channelMap).length > 0
      const comboLabel = msgs.get('combo.subscription.choose', locale)

      Object.values(channelMap).forEach(({ chnl, splitChn, subchannels }) => {
        const hasSubchannels = subchannels.length > 0
        let channelLabel = splitChn && splitChn[2] ? splitChn[2] : 'unknown'
        if (channelLabel === '__ALL__') {
          channelLabel = msgs.get('combo.subscription.all')
        }
        displayChannels.push({
          label: channelLabel,
          chn: chnl,
          splitChn,
          hasSubchannels,
          subchannels
        })
      })
      let selectedIdx =
        displayChannels.length === 1
          ? 0
          : displayChannels.findIndex(({ chn }) => chn === activeChannel)
      if (selectedIdx < 0) {
        selectedIdx = displayChannels.findIndex(({ chn }) => !!chn)
      }

      const hasSubchannelsList = displayChannels[selectedIdx].hasSubchannels
      const channelsLength = hasSubchannelsList
        ? displayChannels[selectedIdx].subchannels.length
        : 0

      const selectedChannelIndex = activeChannel
        ? allChannels.indexOf(activeChannel) + 1
        : 1
      const back1 = '<<'
      const back2 = '<'
      const fwd1 = '>'
      const fwd2 = '>>'

      return (
        // show subscription names only when more than one
        <div className="channel-controls-container">
          {showMainChannel && (
            <div className="channels">
              <div className="subscription label">
                {msgs.get('combo.subscription')}
              </div>

              <div className="channelsCombo">
                <DropdownV2
                  items={displayChannels}
                  id="comboChannel"
                  label={comboLabel}
                  ariaLabel={comboLabel}
                  inline={true}
                  onChange={this.handleSubscriptionChange}
                  selectedItem={displayChannels[selectedIdx]}
                />
              </div>
            </div>
          )}
          {hasSubchannelsList && (
            <div className="pagination">
              <div className="resourcePaging label">
                {msgs.get('subscription.page.label')}
              </div>
              <div className="mainPagination">
                <span
                  id="p1"
                  role="button"
                  className="label pageLabel labelLink"
                  onClick={this.handlePageClick}
                  onKeyPress={this.handlePageClick}
                  tabIndex="0"
                >
                  {back1}
                </span>
                <span
                  id="p2"
                  role="button"
                  className="label pageLabel labelLink"
                  onClick={this.handlePageClick}
                  onKeyPress={this.handlePageClick}
                  tabIndex="0"
                >
                  {back2}
                </span>
                <input
                  className="label pageInput"
                  id="valuePage"
                  onChange={this.handlePageChanged}
                  onKeyPress={this.handlePageClick}
                  aria-label="Current page"
                  type="number"
                  min="1"
                  max="{channelsLength}"
                  value={selectedChannelIndex}
                  tabIndex="0"
                />
                <span className="label pageLabel">
                  {msgs.get('subscription.page.nb', [channelsLength])}
                </span>
                <span
                  id="p3"
                  role="button"
                  className="label pageLabel labelLink"
                  onClick={this.handlePageClick}
                  onKeyPress={this.handlePageClick}
                  tabIndex="0"
                >
                  {fwd1}
                </span>
                <span
                  id="p4"
                  role="button"
                  className="label pageLabel labelLink"
                  onClick={this.handlePageClick}
                  onKeyPress={this.handlePageClick}
                  tabIndex="0"
                >
                  {fwd2}
                </span>
              </div>
            </div>
          )}
        </div>
      )
    }
    return null
  }

  renderView({ style, ...props }) {
    style.marginBottom = -17
    return <div {...props} style={{ ...style }} />
  }

  renderThumbHorizontal() {
    return <div />
  }

  renderThumbVertical({ style, ...props }) {
    const finalStyle = {
      ...style,
      cursor: 'pointer',
      borderRadius: 'inherit',
      backgroundColor: 'rgba(0,0,0,.2)'
    }
    return (
      <div
        className={'channel-controls-scrollbar'}
        style={finalStyle}
        {...props}
      />
    )
  }

  changeSubscriptionChannels(fetchChannel) {
    const { channelControl = {} } = this.props
    const { changeTheChannel } = channelControl

    if (changeTheChannel) {
      this.setState({ fetchChannel })
      changeTheChannel(fetchChannel)
    }
  }
}

export default ChannelControl
