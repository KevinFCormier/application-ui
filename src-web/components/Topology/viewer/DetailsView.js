/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import _ from 'lodash'
import R from 'ramda'
import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { Scrollbars } from 'react-custom-scrollbars'
import {
  Icon,
  NumberInput,
  MultiSelect,
  Button
} from 'carbon-components-react'
import jsYaml from 'js-yaml'
import '../../../../graphics/diagramShapes.svg'
import msgs from '../../../../nls/platform.properties'

const DetailsViewDecorator = ({ shape, className }) => {
  return (
    <svg width="48px" height="48px" viewBox="0 0 48 48">
      <use
        href={`#diagramShapes_${shape}`}
        className={`${className} detailsIcon`}
      />
    </svg>
  )
}

DetailsViewDecorator.propTypes = {
  className: PropTypes.string,
  shape: PropTypes.string
}

class DetailsView extends React.Component {
  constructor(props) {
    super(props)
  }

  handleClick(value) {
    this.processActionLink(value)
  }

  handleKeyPress(value, e) {
    if (e.key === 'Enter') {
      this.processActionLink(value)
    }
  }

  processActionLink(value) {
    const { processActionLink } = this.props
    const { data } = value
    processActionLink(data)
  }

  render() {
    const {
      locale,
      onClose,
      getLayoutNodes,
      getViewContainer,
      staticResourceData,
      selectedNodeId
    } = this.props
    const { typeToShapeMap, getNodeDetails } = staticResourceData
    const currentNode =
      getLayoutNodes().find(n => n.uid === selectedNodeId) || {}
    const { layout = {} } = currentNode
    const resourceType = layout.type || currentNode.type
    const { shape = 'other', className = 'default' } =
      typeToShapeMap[resourceType] || {}
    const details = getNodeDetails(currentNode, locale)
    const name = currentNode.name
    const height = getViewContainer().getBoundingClientRect().height
    const scrollHeight = height * 0.75
    return (
      <section className="topologyDetails">
        <div className="detailsHeader">
          <DetailsViewDecorator shape={shape} className={className} />
          <div>
            <div className="sectionContent">
              <span className="label">{_.capitalize(resourceType)}</span>
            </div>
            <div className="sectionContent">
              <span className="titleNameText">{name}</span>
            </div>
          </div>
          <Icon
            className="closeIcon"
            description={msgs.get('topology.details.close', locale)}
            name="icon--close"
            onClick={onClose}
          />
        </div>
        <hr />
        <Scrollbars
          style={{ height: scrollHeight }}
          renderView={this.renderView}
          renderThumbVertical={this.renderThumbVertical}
          className="details-view-container"
        >
          {details.map(detail => this.renderDetail(detail, locale))}
        </Scrollbars>
      </section>
    )
  }

  renderDetail(detail, locale) {
    switch (detail.type) {
    case 'spacer':
      return this.renderSpacer()
    case 'link':
      return this.renderLink(detail, true)
    case 'number':
      return this.renderNumber(detail, locale)
    case 'selector':
      return this.renderSelector(detail, locale)
    case 'submit':
      return this.renderSubmit(detail, locale)
    case 'snippet':
      return this.renderSnippet(detail, locale)
    default:
      return this.renderLabel(detail, locale)
    }
  }

  renderLabel({ labelKey, labelValue, value, indent, status }, locale) {
    let label = labelValue
    if (labelKey) {
      label = labelValue
        ? msgs.get(labelKey, [labelValue], locale)
        : msgs.get(labelKey, locale)
    }
    label = value ? `${label}:` : label

    const mainSectionClasses = classNames({
      sectionContent: true,
      borderLeft: value !== undefined ? true : false
    })
    const labelClass = classNames({
      label: true,
      sectionLabel: value ? true : false
    })

    const statusIcon = status ? `icon--${status}` : undefined
    return (
      <div className={mainSectionClasses} key={Math.random()}>
        {(labelKey || labelValue) && statusIcon ? (
          <span className="label sectionLabel">
            <Icon
              name={statusIcon}
              fill="#6089bf"
              description=""
              className="label-icon"
            />
            {label}
          </span>
        ) : (
          <span className={labelClass}>{label} </span>
        )}
        {indent && <span className="indent" />}
        <span className="value">{value}</span>
      </div>
    )
  }

  renderSnippet({ value }) {
    if (value) {
      const yaml = jsYaml.safeDump(value).split('\n')
      return (
        <div className="sectionContent snippet">
          {yaml.map(line => {
            return <code key={Math.random()}>{line}</code>
          })}
        </div>
      )
    }
    return null
  }

  renderLink({ value, indent }) {
    const handleClick = this.handleClick.bind(this, value)
    const handleKeyPress = this.handleKeyPress.bind(this, value)
    const showLaunchOutIcon = !R.pathOr(false, ['data', 'specs', 'isDesign'])(
      value
    ) //if not show yaml

    let iconName = 'icon--arrow--right'
    if (_.get(value, 'label', '').startsWith('http')) {
      iconName = 'icon--launch'
    }
    const mainSectionClasses = classNames({
      sectionContent: true,
      borderLeft: indent ? true : false
    })

    const linkLabelClasses = classNames({
      link: true,
      sectionLabel: indent ? true : false
    })

    return (
      <div className={mainSectionClasses} key={Math.random()}>
        <span
          className={linkLabelClasses}
          id="linkForNodeAction"
          tabIndex="0"
          role={'button'}
          onClick={handleClick}
          onKeyPress={handleKeyPress}
        >
          {value.label}
          {showLaunchOutIcon && (
            <Icon
              name={iconName}
              fill="#6089bf"
              description=""
              className="label-icon"
            />
          )}
        </span>
      </div>
    )
  }

  renderNumber({ labelKey, value }, locale) {
    return (
      <div className="sectionContent form" key={Math.random()}>
        {labelKey && (
          <span className="label">{msgs.get(labelKey, locale)}: </span>
        )}
        <NumberInput
          id={labelKey}
          label={''}
          value={value}
          min={1}
          max={10}
          step={1}
          onChange={this.enableSubmitBtn.bind(this)}
        />
      </div>
    )
  }

  renderSelector({ labelKey, value, other }, locale) {
    return (
      <div className="sectionContent form" key={Math.random()}>
        {labelKey && (
          <span className="label">{msgs.get(labelKey, locale)}: </span>
        )}
        <MultiSelect.Filterable
          items={other}
          initialSelectedItems={value}
          placeholder={value.join(', ')}
          itemToString={item => item}
          onChange={this.enableSubmitBtn.bind(this)}
        />
      </div>
    )
  }

  setSubmitBtn = ref => {
    this.submitBtn = ref
  };
  renderSubmit() {
    return (
      <div className="sectionContent submit" key={Math.random()}>
        <Button
          small
          id="edit-button"
          disabled
          ref={this.setSubmitBtn}
          onClick={this.onSubmit}
        >
          {msgs.get('modal.button.submit', this.props.locale)}
        </Button>
      </div>
    )
  }

  enableSubmitBtn() {
    this.submitBtn.disabled = false
  }

  onSubmit() {
    this.props.onClose()
  }

  renderSpacer() {
    return (
      <div className="sectionContent" key={Math.random()}>
        <div className="spacer" />
      </div>
    )
  }

  renderView({ style, ...props }) {
    style.marginBottom = -17
    return <div {...props} style={{ ...style }} />
  }

  renderThumbVertical({ style, ...props }) {
    const finalStyle = {
      ...style,
      cursor: 'pointer',
      borderRadius: 'inherit',
      backgroundColor: 'rgba(0,0,0,.2)'
    }
    return (
      <div className={'details-view-scrollbar'} style={finalStyle} {...props} />
    )
  }
}

DetailsView.propTypes = {
  getLayoutNodes: PropTypes.func,
  getViewContainer: PropTypes.func,
  locale: PropTypes.string,
  onClose: PropTypes.func,
  processActionLink: PropTypes.func,
  selectedNodeId: PropTypes.string,
  staticResourceData: PropTypes.object
}

export default DetailsView
