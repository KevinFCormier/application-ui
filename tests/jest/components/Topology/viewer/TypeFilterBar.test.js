/*******************************************************************************
 * Copyright (c) 2020 Red Hat, Inc.
 *******************************************************************************/
"use strict";

import { FilterButton } from "../../../../../src-web/components/Topology/viewer/TypeFilterBar";
import renderer from "react-test-renderer";
import React from "react";

describe("FilterButton basic render", () => {
  const mockData = {
    type: "application",
    displayName: "application",
    selected: false,
    toolip: "application",
    typeToShapeMap: {
      application: {
        shape: "application",
        className: "design",
        nodeRadius: 30
      },
      deployable: { shape: "deployable", className: "design" },
      subscription: { shape: "subscription", className: "design" },
      rules: { shape: "rules", className: "design" },
      clusters: { shape: "cluster", className: "container" },
      helmrelease: { shape: "chart", className: "container" },
      package: { shape: "chart", className: "container" },
      internet: { shape: "cloud", className: "internet" },
      host: { shape: "host", className: "host" },
      policy: { shape: "roundedSq", className: "design", nodeRadius: 30 },
      placement: { shape: "placement", className: "design" },
      cluster: { shape: "cluster", className: "container" },
      service: { shape: "service", className: "service" },
      deployment: { shape: "deployment", className: "deployment" },
      daemonset: { shape: "star4", className: "daemonset" },
      statefulset: { shape: "cylinder", className: "statefulset" },
      pod: { shape: "pod", className: "pod" },
      container: { shape: "irregularHexagon", className: "container" },
      cronjob: { shape: "clock", className: "default" },
      spare1: { shape: "star4", className: "daemonset" },
      spare2: { shape: "roundedSq", className: "daemonset" },
      spare3: { shape: "hexagon", className: "daemonset" },
      spare4: { shape: "irregularHexagon", className: "daemonset" },
      spare5: { shape: "roundedRect", className: "daemonset" }
    },
    handleClickFunc: jest.fn()
  };
  it("renders as expected", () => {
    const component = renderer.create(
      <FilterButton
        key={mockData.type}
        label={mockData.displayName}
        selected={mockData.selected}
        typeToShapeMap={mockData.typeToShapeMap}
        tooltip={mockData.tooltip}
        handleClick={mockData.handleClickFunc}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
