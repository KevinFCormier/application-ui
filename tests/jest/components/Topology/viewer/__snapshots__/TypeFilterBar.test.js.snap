// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterButton basic render renders as expected 1`] = `
<div
  className="filter-bar-button-container"
  title=""
>
  <div
    aria-checked={false}
    className="filter-bar-button"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="checkbox"
    tabIndex="0"
  >
    <svg
      height="16px"
      width="16px"
    >
      <use
        className="design filter-bar-button-icon"
        href="#diagramShapes_application"
      />
    </svg>
    <div
      className="filter-bar-button-label"
    >
      application
    </div>
  </div>
</div>
`;
